fuel/datasets/hdf5.py

```diff
@@ -64,3 +67,124 @@ def get_data(self, state=None, request=None):
                 raise ValueError
         data = [node[request] for node in self.nodes]
         return data
+
+
+@do_not_pickle_attributes('data_sources')
+class H5PYDataset(Dataset):
+    """An h5py-fueled HDF5 dataset.
+
+    This dataset class assumes a particular file layout:
+
+    * Data sources reside in the root group, and their names define the
+      source names.
+    * The dataset is not explicitly split. Instead, splits are defined as
+      attributes of the root group. They're expected to be numpy arrays of
+      shape (2,), with the first element being the starting point
+      (inclusive) of the split and the last element being the stopping
+      point (exclusive) of the split.
+
+    The `which_set`, `start` and `stop` parameters work together in the
+    following way:
+
+    * `which_set` is resolved first. If it is `None`, the whole dataset is
+      used.
+    * `start` and `stop` define a slice *within the context of*
+      `which_set`.
+
+    Parameters
+    ----------
+    path : str
+        Path to the HDF5 file.
+    which_set : str, optional
+        Name of the root group attribute containing the split information.
+        Defaults to `None`, in which case the whole dataset is used.
+    subset : slice, optional
+        A slice of data *within the context of the split* to use. Defaults
+        to `None`, in which case the whole split is used. **Note:
+        at the moment, `slice.step` must be either 1 or `None`.**
+    load_in_memory : bool, optional
+        Whether to load the data in main memory. Defaults to `False`.
+    driver : str, optional
+        Low-level driver to use. Defaults to `None`. See h5py
+        documentation for a complete list of available options.
+
+    """
+    ref_counts = dict()
+
+    def __init__(self, path, which_set=None, subset=None, load_in_memory=False,
+                 driver=None, **kwargs):
+        self.path = path
+        self.which_set = which_set
+        self.subset = subset if subset else slice(None, None, None)
+        self.load_in_memory = load_in_memory
+        self.driver = driver
+        self.load()
+        super(H5PYDataset, self).__init__(**kwargs)
+
+    def _get_file_id(self):
+        try:
+            return next(
+                filter(lambda x: x.name == self.path, self.ref_counts.keys()))
```

Comment:
And instead of `try` and `except` maybe
```python
file_id = [f.name for f in self.ref_counts.keys()]
if not file_id:
    return self.path
file_id, = file_id  # Raises an error if there are multiple matches, which we probably want?
return file_id
```